pennylane/gradients/general_shift_rules.py

```diff
@@ -126,4 +166,152 @@ def get_shift_rule(frequencies, shifts=None):
 
     coeffs = np.concatenate((coeffs, -coeffs))
     shifts = np.concatenate((shifts, -shifts))
-    return ([[coeffs[mu], 1, shifts[mu]] for mu in range(0, 2 * n_freqs)],)
+    return np.stack([coeffs, shifts])
+
+
+@functools.lru_cache()
+def generate_shift_rule(frequencies, shifts=None, order=1):
+    r"""Computes the parameter shift rule for a unitary based on its generator's eigenvalue frequency
+    spectrum.
+
+    To compute gradients of circuit parameters in variational quantum algorithms, expressions for
+    cost function first derivatives with respect to the variational parameters can be cast into
+    linear combinations of expectation values at shifted parameter values. These "gradient recipes"
+    can be obtained from the unitary generator's eigenvalue frequency spectrum. Details can be
+    found in https://arxiv.org/abs/2107.12390.
+
+    Args:
+        frequencies (tuple[int or float]): The tuple of eigenvalue frequencies. Eigenvalue
+            frequencies are defined as the unique positive differences obtained from a set of
+            eigenvalues.
+        shifts (tuple[int or float]): the tuple of shift values. If unspecified, equidistant
+            shifts are assumed. If supplied, the length of this tuple should match the number of
+            given frequencies.
+        order (int): the order of differentiation to compute the shift rule for
+
+    Returns:
+        tuple: a tuple of coefficients and shifts describing the gradient recipe
+        for the parameter-shift method. For parameter :math:`\phi`, the
+        coefficients :math:`c_i` and the shifts :math:`s_i` combine to give a gradient
+        recipe of the following form:
+
+        .. math:: \frac{\partial}{\partial\phi}f = \sum_{i} c_i f(\phi + s_i).
+
+        where :math:`f(\phi) = \langle 0|U(\phi)^\dagger \hat{O} U(\phi)|0\rangle`

```

Comment:
I'm not sure I fully follow?

Reply:
For an input of `R` frequencies and `R` shifts, the output will have shape `(2, 2R)` and not `(2, R)`. That is, every user-given shift value will be used twice. Maybe this could lead to surprises in the number of executions used? But maybe this is not a surprise that ever happens and I'm overthinking here..